vm/starnix/starnix.go

```diff
@@ -0,0 +1,383 @@
+		env:   env,
+		cfg:   cfg,
+	}
+	return pool, nil
+}
+
+func (pool *Pool) Count() int {
+	return pool.count
+}
+
+func (pool *Pool) Create(workdir string, index int) (vmimpl.Instance, error) {
+	inst := &instance{
+		ffxBinary:        ".jiri_root/bin/ffx",
+		fuchsiaDirectory: pool.env.KernelSrc,
+		name:             fmt.Sprintf("VM-%v", index),
+		index:            index,
+		cfg:              pool.cfg,
+		debug:            pool.env.Debug,
+		timeouts:         pool.env.Timeouts,
+		workdir:          workdir,
+		// This file is auto-generated inside createAdbScript.
+		executor:  filepath.Join(workdir, "adb_executor.sh"),
+		targetDir: "/data",
+	}
+	closeInst := inst
+	defer func() {
+		if closeInst != nil {
+			closeInst.Close()
+		}
+	}()
+
+	var err error
+	inst.rpipe, inst.wpipe, err = osutil.LongPipe()
+	if err != nil {
+		return nil, err
+	}
+
+	if err := inst.boot(); err != nil {
+		return nil, err
+	}
+
+	closeInst = nil
+	return inst, nil
+}
+
+func (inst *instance) terminateFuchsiaVM() {
+	if inst.debug {
+		log.Logf(0, "terminating Fuchsia VM: %v", inst.name)
+	}
+	inst.runCommand(1*time.Minute*inst.timeouts.Scale, inst.ffxBinary, "emu", "stop", inst.name)
+}
+
+func (inst *instance) Close() {
+	if inst.fuchsiaVMStarted {
+		inst.terminateFuchsiaVM()
+	}
+	if inst.fuchsiaLogs != nil {
+		inst.fuchsiaLogs.Process.Kill()
+		inst.fuchsiaLogs.Wait()
+	}
+	if inst.adb != nil {
+		inst.adb.Process.Kill()
+		inst.adb.Wait()
+	}
+	if inst.merger != nil {
+		inst.merger.Wait()
+	}
+	if inst.rpipe != nil {
+		inst.rpipe.Close()
+	}
+	if inst.wpipe != nil {
+		inst.wpipe.Close()
+	}
+}
+
+func (inst *instance) startFuchsiaVM() error {
+	ffxArgs := []string{"emu", "start", "--headless", "--name", inst.name}
+	inst.args = ffxArgs[2:]
+	err := inst.runCommand(5*time.Minute*inst.timeouts.Scale, inst.ffxBinary, ffxArgs...)
+	if err != nil {
+		return err
+	}
+	inst.fuchsiaVMStarted = true
+	return nil
+}
+
+func (inst *instance) startFuchsiaLogs() (*exec.Cmd, error) {
+	cmd := osutil.Command(inst.ffxBinary, "--target", inst.name, "log")
+	cmd.Dir = inst.fuchsiaDirectory
+	cmd.Stdout = inst.wpipe
+	cmd.Stderr = inst.wpipe
+	inst.merger.Add("fuchsia", inst.rpipe)
+	return cmd, cmd.Start()
+}
+
+func (inst *instance) startAdbServerAndConnection(timeout time.Duration) error {
+	cmd := osutil.Command(inst.ffxBinary, "--target", inst.name, "starnix", "adb", "-p", fmt.Sprintf("%d", inst.port))
+	cmd.Dir = inst.fuchsiaDirectory
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+	if inst.debug {
+		log.Logf(0, fmt.Sprintf("the adb bridge is listening on 127.0.0.1:%d", inst.port))
+	}
+	inst.adb = cmd
+	return inst.connectToAdb(timeout)
+}
+
+func (inst *instance) connectToAdb(timeout time.Duration) error {
+	startTime := time.Now()
+	for {
+		vmimpl.SleepInterruptible(3 * time.Second)
+		if inst.debug {
+			log.Logf(0, "attempting to connect to ADB")
+		}
+		cmd := osutil.Command("adb", "connect", fmt.Sprintf("127.0.0.1:%d", inst.port))
```

Comment:
Does adb come from the fuchsia checkout? If so use full path. Or users are supposed to pre-install it?

Reply:
Adb should be installed by users, adding a check in the vmimpl.Env ctor
```
        if _, err := exec.LookPath("adb"); err != nil {
                return nil, err
        }
```